drivers/gpio/gpio_gd32.c

```diff
@@ -0,0 +1,369 @@
+
+#ifdef CONFIG_GD32_HAS_AF_PINMUX
+	GPIO_TG(config->reg) = pins;
+#else
+	GPIO_OCTL(config->reg) ^= pins;
+#endif /* CONFIG_GD32_HAS_AF_PINMUX */
+
+	return 0;
+}
+
+static int gpio_gd32_pin_interrupt_configure(const struct device *port,
+					     gpio_pin_t pin,
+					     enum gpio_int_mode mode,
+					     enum gpio_int_trig trig)
+{
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		gd32_exti_disable(pin);
+		(void)gd32_exti_configure(pin, NULL, NULL);
+		gd32_exti_trigger(pin, GD32_EXTI_TRIG_NONE);
+	} else if (mode == GPIO_INT_MODE_EDGE) {
+		int ret;
+
+		ret = gpio_gd32_configure_extiss(port, pin);
+		if (ret < 0) {
+			return ret;
+		}
+
+		switch (trig) {
+		case GPIO_INT_TRIG_LOW:
+			gd32_exti_trigger(pin, GD32_EXTI_TRIG_FALLING);
+			break;
+		case GPIO_INT_TRIG_HIGH:
+			gd32_exti_trigger(pin, GD32_EXTI_TRIG_RISING);
+			break;
+		case GPIO_INT_TRIG_BOTH:
+			gd32_exti_trigger(pin, GD32_EXTI_TRIG_BOTH);
+			break;
+		default:
+			return -ENOTSUP;
+		}
+
+		ret = gd32_exti_configure(pin, gpio_gd32_isr, (void *)port);
+		if (ret < 0) {
+			return ret;
+		}
+
+		gd32_exti_enable(pin);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int gpio_gd32_manage_callback(const struct device *dev,
+				     struct gpio_callback *callback, bool set)
+{
+	struct gpio_gd32_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static const struct gpio_driver_api gpio_gd32_api = {
+	.pin_configure = gpio_gd32_configure,
+	.port_get_raw = gpio_gd32_port_get_raw,
+	.port_set_masked_raw = gpio_gd32_port_set_masked_raw,
+	.port_set_bits_raw = gpio_gd32_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_gd32_port_clear_bits_raw,
+	.port_toggle_bits = gpio_gd32_port_toggle_bits,
+	.pin_interrupt_configure = gpio_gd32_pin_interrupt_configure,
+	.manage_callback = gpio_gd32_manage_callback,
+};
+
+static int gpio_gd32_init(const struct device *port)
+{
+	const struct gpio_gd32_config *config = port->config;
+
+	rcu_periph_clock_enable(config->rcu_periph_clock);
+
+#ifdef CONFIG_GD32_HAS_AF_PINMUX
+	/* enable access to SYSCFG_EXTISSn registers */
+	rcu_periph_clock_enable(DT_PROP(SYSCFG_NODE, rcu_periph_clock));
+#else
+	/* enable access to AFIO_EXTISSn registers */
+	rcu_periph_clock_enable(DT_PROP(AFIO_NODE, rcu_periph_clock));
+#endif /* CONFIG_GD32_HAS_AF_PINMUX */
+
+	return 0;
+}
+
+#define GPIO_GD32_DEFINE(n)						       \
+	static const struct gpio_gd32_config gpio_gd32_config##n = {	       \
+		.common = {						       \
+			.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(n),   \
+		},							       \
+		.reg = DT_INST_REG_ADDR(n),				       \
+		.rcu_periph_clock = DT_INST_PROP(n, rcu_periph_clock),	       \
+	};								       \
+									       \
+	static struct gpio_gd32_data gpio_gd32_data##n;			       \
+									       \
+	DEVICE_DT_INST_DEFINE(n, &gpio_gd32_init, NULL, &gpio_gd32_data##n,    \
+			      &gpio_gd32_config##n, PRE_KERNEL_1,	       \
```

Comment:
GPIO driver initialize at PRE_KERNEL_1 level. Others driver, like spi and adc, normally initialize at POST_KERNEL level.
when they call pinctrl_apply_state function to adjust pin function, there may have a risk to incorrectly disable the GPIO.

https://github.com/zephyrproject-rtos/zephyr/blob/521e6ffb7c1fcb663b98546eeba7822e03ca5a49/drivers/pinctrl/pinctrl_gd32_af.c#L98

Reply:
Fixed, thanks for spotting. We can likely improve this once a proper clock control API is in place.